plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/ui/contentAssist/ContentAssistFragment2.xtend

```diff
@@ -0,0 +1,330 @@
+			if (generateStub) grammar.getProposalProviderClass else grammar.getGenProposalProviderClass;
+		
+		if (projectConfig.eclipsePluginManifest != null) {
+			projectConfig.eclipsePluginManifest.requiredBundles += "org.eclipse.xtext.ui"
+		}
+
+		new GuiceModuleAccess.BindingFactory()
+				.addTypeToType(
+					new TypeReference("org.eclipse.xtext.ui.editor.contentassist.IContentProposalProvider"),
+					chosenClass
+				).contributeTo(language.eclipsePluginGenModule);
+
+		var generatedCode = false
+
+		if (projectConfig.eclipsePluginSrcGen !== null) {
+			// generate the 'Abstract...ProposalProvider'
+			generateGenJavaProposalProvider
+			generatedCode = true
+		}
+
+		if (generateStub && projectConfig.eclipsePluginSrc != null) {
+			if (preferXtendStubs) {
+				generateXtendProposalProviderStub
+
+				if (projectConfig.eclipsePluginManifest != null) {
+					projectConfig.eclipsePluginManifest.requiredBundles += "org.eclipse.xtext.xbase.lib"
+				}
+			} else {
+				generateJavaProposalProviderStub
+			}
+
+			generatedCode = true
+		}
+
+		if (generatedCode && projectConfig.eclipsePluginManifest != null) {
+			projectConfig.eclipsePluginManifest.exportedPackages += grammar.proposalProviderClass.packageName
+		}
+	}
+
+	// generation of the '...ProposalProvider' stub
+
+	def generateXtendProposalProviderStub() {
+		fileAccessFactory.createXtendFile(grammar.proposalProviderClass, '''
+			/**
+			 * See https://www.eclipse.org/Xtext/documentation/304_ide_concepts.html#content-assist
+			 * on how to customize the content assistant.
+			 */
+			class �grammar.proposalProviderClass.simpleName� extends �grammar.genProposalProviderClass� {
+			}
+		''').writeTo(projectConfig.eclipsePluginSrc)
+	}
+
+	def generateJavaProposalProviderStub() {
+		fileAccessFactory.createJavaFile(grammar.proposalProviderClass, '''
+			/**
+			 * See https://www.eclipse.org/Xtext/documentation/304_ide_concepts.html#content-assist
+			 * on how to customize the content assistant.
+			 */
+			public class �grammar.proposalProviderClass.simpleName� extends �grammar.genProposalProviderClass� {
+			}
+		''').writeTo(projectConfig.eclipsePluginSrc)
+	}
+
+	// generation of the 'Abstract...ProposalProvider'
+
+	
+	def generateGenJavaProposalProvider() {
+		// excluded features are those that stem from inherited grammars,
+		//  they are handled by the super grammars' proposal provider
+		val excludedFqnFeatureNames = grammar.getFQFeatureNamesToExclude
+		val processedNames = newHashSet()
+
+		// determine all assignments within the grammar that are not excluded
+		//  and not handled yet, (iterable is evaluated lazily!)
+		val assignments = grammar.containedAssignments().fold(<Assignment>newArrayList()) [ candidates, assignment |
+			val fqFeatureName = assignment.FQFeatureName
+			if (!processedNames.contains(fqFeatureName) && !excludedFqnFeatureNames.contains(fqFeatureName)) {
+				processedNames += fqFeatureName;
+				candidates += assignment;
+			}
+			candidates
+		]
```

Comment:
I'd rather evaluate this eagerly to stay with the imperative style. Otherwise this will break if someone refactors the template further down, which would be very confusing.